File: legacy/pasarguardbot/frontend/src/types/panel/marketing.ts
/** Mirrors app/models/panel/marketing.py */
import type { PagedRequest, PageMeta, PanelAuthRequest, PanelEnvelope } from "./common";

export const DISCOUNT_CODE_PATTERN = /^[A-Za-z0-9_-]{2,40}$/;

export interface PanelDiscountRow {
  id: number;
  code: string;
  discount_percentage: number;
  usage_limit: number;
  times_used: number;
  expiration_date?: number | null;
  user_id?: number | null;
  is_public: boolean;
  expired: boolean;
  exhausted: boolean;
}

export type PanelDiscountsRequest = PagedRequest;

export interface PanelDiscountsResponse extends PanelEnvelope {
  discounts: PanelDiscountRow[];
  meta: PageMeta;
}

export interface PanelDiscountSaveRequest extends PanelAuthRequest {
  code_id?: number | null;
  code: string;
  discount_percentage: number;
  usage_limit?: number;
  /** Null means no expiry. */
  expires_days?: number | null;
  /** Null means the code is public. */
  user_id?: number | null;
  is_public?: boolean;
}

export interface PanelDiscountDeleteRequest extends PanelAuthRequest {
  code_id: number;
}

export interface PanelReferralSettings {
  referral_enabled: boolean;
  referral_reward_amount: number;
  referral_bonus_amount: number;
  referral_banner_text?: string | null;
}

export interface PanelReferralRewardRow {
  id: number;
  referrer_id?: number | null;
  referred_id?: number | null;
  reward_amount: number;
  bonus_amount: number;
  status?: string | null;
  created_at?: number | null;
}

export type PanelReferralRequest = PagedRequest;

export interface PanelReferralResponse extends PanelEnvelope {
  settings: PanelReferralSettings;
  rewards: PanelReferralRewardRow[];
  meta: PageMeta;
  total_rewarded: number;
  total_paid: number;
  total_bonus: number;
}

export interface PanelReferralSaveRequest extends PanelAuthRequest {
  referral_enabled: boolean;
  referral_reward_amount: number;
  referral_bonus_amount: number;
  referral_banner_text?: string;
}

File: legacy/pasarguardbot/app/models/panel/marketing.py
"""Admin panel DTOs: discount codes and the referral programme."""

from pydantic import BaseModel, Field

from app.models.panel.common import PagedRequest, PageMeta, PanelRequest, PanelResponse

CODE_PATTERN = r"^[A-Za-z0-9_\-]{2,40}$"


class PanelDiscountRow(BaseModel):
    id: int
    code: str
    discount_percentage: int = 0
    usage_limit: int = 0
    times_used: int = 0
    expiration_date: int | None = None
    user_id: int | None = None
    is_public: bool = True
    expired: bool = False
    exhausted: bool = False


class PanelDiscountsRequest(PagedRequest):
    pass


class PanelDiscountsResponse(PanelResponse):
    discounts: list[PanelDiscountRow] = Field(default_factory=list)
    meta: PageMeta = Field(default_factory=PageMeta)


class PanelDiscountSaveRequest(PanelRequest):
    """Create when ``code_id`` is null, otherwise update that code."""

    code_id: int | None = None
    code: str = Field(..., pattern=CODE_PATTERN)
    discount_percentage: int = Field(..., ge=1, le=100)
    usage_limit: int = Field(1, ge=1)
    expires_days: int | None = Field(None, ge=0, description="Null means no expiry")
    user_id: int | None = Field(None, description="Null means the code is public")
    is_public: bool = True


class PanelDiscountDeleteRequest(PanelRequest):
    code_id: int


class PanelReferralSettings(BaseModel):
    referral_enabled: bool = True
    referral_reward_amount: int = 0
    referral_bonus_amount: int = 0
    referral_banner_text: str | None = None


class PanelReferralRewardRow(BaseModel):
    id: int
    referrer_id: int | None = None
    referred_id: int | None = None
    reward_amount: int = 0
    bonus_amount: int = 0
    status: str | None = None
    created_at: int | None = None


class PanelReferralRequest(PagedRequest):
    pass


class PanelReferralResponse(PanelResponse):
    settings: PanelReferralSettings = Field(default_factory=PanelReferralSettings)
    rewards: list[PanelReferralRewardRow] = Field(default_factory=list)
    meta: PageMeta = Field(default_factory=PageMeta)
    total_rewarded: int = 0
    total_paid: int = 0
    total_bonus: int = 0


class PanelReferralSaveRequest(PanelRequest):
    referral_enabled: bool = True
    referral_reward_amount: int = Field(0, ge=0)
    referral_bonus_amount: int = Field(0, ge=0)
    referral_banner_text: str = Field("", max_length=4096)
